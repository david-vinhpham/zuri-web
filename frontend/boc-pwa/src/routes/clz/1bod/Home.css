/**
 * Home Credit (https://www.homecredit.vn/)
 *
 * Copyright Â© 2018-present HomeCredit, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

@import '../../../components/variables.css';
@import './grid.css';

.modalFooter {
  padding: var(--b-space-3xs) var(--b-space-md);
  background-color: var(--c-brand);
  color: white;
  text-align: center;
}

.errorText {
  text-align: center;
  color: var(--c-brand);
}

.modalBody {
  text-align: center;
  color: var(--c-text);
  font-size: var(--f-text);
  font-style: italic;
  padding-bottom: var(--b-space-md);

  & h3 {
    font-style: normal;
  }

  & p {
    text-align: left;
  }
}

.modalBody .modalText {
  text-align: center;
}

.headerbar {
  position: relative;
  z-index: var(--z-index-high);
}

.callBtn {
  position: fixed;
  bottom: var(--b-space-xl);
  right: var(--b-space-xl);
  z-index: var(--z-index-higher);
}

@media (--lg) {
  .homeWrapper .loading {
    position: fixed;
    top: 0;
    left: 0;
    height: 100vh;
    width: 100vw;
  }
}

@media (--sm) {
  .layoutWrapper {
    display: flex;
    flex-direction: column;
    justify-content: space-between;
  }

  .callBtn {
    bottom: var(--b-space-2xs);
    right: var(--b-space-2xs);
  }

  .callBtnSpec {
    right: var(--b-space-3xs);
    bottom: var(--b-space-4xl);
  }
}
