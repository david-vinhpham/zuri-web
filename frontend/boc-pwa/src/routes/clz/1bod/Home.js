/**
 * BOC VN (http://www.bocvietnam.com/)
 *
 * Copyright © 2018-present BOCVN, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import { func, number, shape, string, objectOf, any, bool } from 'prop-types';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { get, noop, isString } from 'lodash';
import withStyles from 'isomorphic-style-loader/lib/withStyles';

import HeaderBar from 'components/Header/HeaderBar';
import { Modal } from 'homecredit-ui';

import {
  setFormStatus,
  updateStep,
  set1BODInfo,
  trackingInfo,
} from 'actions/clz';
import { verifyShortId } from 'actions/shortURL';

import { FORM_STATUS, MESSAGES } from 'constants/common';
import { convertDate } from 'utils/time';
import classnames from 'classnames';
import FullLoading from 'components/common/FullLoading/FullLoading';
import Cycle from 'components/common/FullLoading/Cycle';

import {
  HEADER_TITLES,
  STEPS,
  STEP_NAME,
  STEP_COMPONENT,
  SUBMIT_APPLICATION_FAIL,
  INVALID_SA_CODE,
  AGREE,
  UNIFIED_ID_TYPE,
} from './constant';
import history from '../../../history';
import CallButton from './CallButton';

import s from './Home.css';
import DesktopDesign from './DesktopDesign';
import NextButton from './NextButton';

const FullLoadingCycle = FullLoading(Cycle);

class FirstBOD extends Component {
  static propTypes = {
    params: shape({
      sa: string,
    }),
    updateFormStatus: func,
    updateStep: func,
    onChange1BODInfo: func,
    step: number,
    trackingToken: string,
    firstBODInfo: objectOf(any),
    loaderStatus: bool,
  };

  static defaultProps = {
    params: {
      sa: null,
    },
    updateFormStatus: noop,
    updateStep: noop,
    onChange1BODInfo: noop,
    step: STEPS.WELCOME_PAGE,
    trackingToken: null,
    firstBODInfo: {},
    loaderStatus: false,
  };

  constructor(props) {
    super(props);
    this.state = {
      isOpenModal: false,
      modalId: '',
    };
  }

  async componentDidMount() {
    const { step: componentStep } = this.props;
    // Save time begin for GA
    this.startTime = Date.now();

    // Parse SA Code
    const {
      params: { step, sa },
      onChange1BODInfo,
    } = this.props;

    let partner;
    let tipper;

    if (step && step !== STEP_NAME[componentStep]) {
      partner = step;
      tipper = sa;
    }

    if (partner || sa) {
      try {
        const verify = await verifyShortId(partner || sa);
        const { partnerCode, saCode } = verify;

        if (partnerCode) {
          onChange1BODInfo('partnerCode', partnerCode);
          onChange1BODInfo('tipperCode', tipper);
          onChange1BODInfo('retailAgent', saCode);
          onChange1BODInfo('sa', null);
        } else if (saCode) {
          onChange1BODInfo('retailAgent', saCode);
          onChange1BODInfo('sa', sa);
          onChange1BODInfo('partnerCode', null);
          onChange1BODInfo('tipperCode', null);
        }

        if (verify.errors.length) {
          history.push(`/${STEP_NAME[STEPS.WELCOME_PAGE]}`);
        }
      } catch (error) {
        history.push(`/${STEP_NAME[STEPS.WELCOME_PAGE]}`);
      }
    }

    if (!step && !sa) {
      history.push(`/${STEP_NAME[STEPS.WELCOME_PAGE]}`);
    }
  }

  /**
   * Reset 1bod step
   *
   * @return {void}
   */
  componentWillUnmount() {
    this.props.updateStep(STEPS.WELCOME_PAGE);

    // Send time begin for GA
    const timeFill = Date.now() - this.startTime;
    window.ga('send', 'timing', {
      timingCategory: 'Application process',
      timingVar: '1BOD application',
      timingValue: timeFill,
    });
  }

  /**
   * Go back to the previous step
   *
   * @return {void}
   */
  onClickBack = () => {
    const { step = STEPS.CALCULATE_CASH } = this.props;
    let previousStep = step - 1;
    // Do not allow to back to step 3
    previousStep === 3 && (previousStep -= 1);
    this.props.updateStep(previousStep);
    history.push(`/${STEP_NAME[previousStep]}`);
  };

  /**
   * Set status to form is need validate
   *
   * @return {void}
   */
  onClickNext = () => {
    this.props.updateFormStatus(FORM_STATUS.NEED_VALIDATE);
  };

  /**
   * Update state when modal is opened
   */
  onOpenModal = (modalId = '') => {
    this.setState({
      isOpenModal: true,
      modalId,
    });
  };

  /**
   * Update state when modal is closed
   */
  onCloseModal = () => {
    this.setState({
      isOpenModal: false,
    });
  };

  /**
   * Go to next step, use in children component
   *
   * @return {void}
   */
  goToNextStep = () => {
    const { step, trackingToken, firstBODInfo } = this.props;
    const nextStep = step + 1;
    this.props.updateStep(nextStep);
    history.push(`/${STEP_NAME[nextStep]}`);
    const trackedInfo = this.resolveDataTracking(firstBODInfo, step);
    const data = {
      url: STEP_NAME[step],
      body: trackedInfo,
      token: trackingToken,
    };
    (step === STEPS.USER_INFO || step === STEPS.SOCIAL_INFO) &&
      trackingInfo(data);
  };

  resolveDataTracking = (data, step) => {
    switch (step) {
      case STEPS.USER_INFO: {
        const { firstName, lastName, middleName, gender } = data;
        const dob = get(data, 'dayOfBirth');
        const dayOfBirth = isString(dob)
          ? convertDate(new Date(dob), 'YYYY-MM-DD')
          : convertDate(dob, 'YYYY-MM-DD');

        return {
          dayOfBirth,
          firstName,
          gender,
          lastName,
          middleName,
        };
      }
      case STEPS.SOCIAL_INFO: {
        const {
          monthlyIncome,
          monthlyPayment,
          identificationType,
          drivingLicence,
          idCard,
          familyBook,
        } = data;
        const rawData = {
          drivingLicence,
          familyBook,
          idCard,
          identificationType,
          monthlyIncome,
          otherMonthlyInstalment: monthlyPayment,
        };

        rawData[UNIFIED_ID_TYPE[identificationType]] = '';

        return rawData;
      }
      default: {
        return {};
      }
    }
  };

  render() {
    const { step = STEPS.WELCOME_PAGE, params, loaderStatus } = this.props;
    const { isOpenModal, modalId } = this.state;
    const ContentComponent = get(STEP_COMPONENT, step);
    const popup = { open: this.onOpenModal, close: this.onCloseModal };
    const isFirstAndLastStep = [0, 7].includes(step);

    return (
      <div className={s.homeWrapper}>
        <DesktopDesign>
          <div className={s.layoutWrapper}>
            <div>
              {!isFirstAndLastStep && (
                <HeaderBar
                  title={get(HEADER_TITLES, step)}
                  onBackClick={this.onClickBack}
                  onNextClick={this.onClickNext}
                  hideNextButton
                  className={s.headerbar}
                />
              )}
              <ContentComponent
                goToNextStep={this.goToNextStep}
                globalPopup={popup}
                params={params}
                onNextClick={this.onClickNext}
              />
            </div>
          </div>
          {loaderStatus && <FullLoadingCycle className={s.loading} />}
          <NextButton
            onNextClick={this.onClickNext}
            hideNextButton={isFirstAndLastStep}
          />
        </DesktopDesign>
        <NextButton
          mobile
          onNextClick={this.onClickNext}
          hideNextButton={isFirstAndLastStep}
        />
        <CallButton
          className={classnames(s.callBtn, step === 0 && s.callBtnSpec)}
        />

        <Modal
          onOpen={this.onOpenModal}
          onClose={this.onCloseModal}
          isOpen={isOpenModal}
          footer={
            <div className={s.modalFooter} onClick={this.onCloseModal}>
              {(!modalId || modalId === INVALID_SA_CODE) && (
                <h4>{MESSAGES.OK}</h4>
              )}

              {modalId === SUBMIT_APPLICATION_FAIL && <h4>{AGREE}</h4>}
            </div>
          }
        >
          {!modalId && <p className={s.errorText}>{MESSAGES.GLOBAL_ERROR}</p>}

          {modalId === SUBMIT_APPLICATION_FAIL && (
            <div className={s.modalBody}>
              <h3>Lưu ý</h3>

              <p>Đơn vay chưa được đăng ký thành công.</p>

              <p>
                Quý khách vui lòng kiểm tra lại thông tin đã nhập, hoặc liên hệ
                nhân viên http://www.bocvietnam.com gần nhất để được hỗ trợ.
              </p>
            </div>
          )}

          {modalId === INVALID_SA_CODE && (
            <div className={s.modalBody}>
              <h3>Lưu ý</h3>
              <p className={s.modalText}>Mã SA không hợp lệ</p>
            </div>
          )}
        </Modal>
      </div>
    );
  }
}

const mapStateToProps = ({ clz }) => ({
  formStatus: clz.formStatus,
  step: clz.step,
  trackingToken: clz.trackingToken,
  firstBODInfo: clz.firstBODInfo,
  loaderStatus: clz.loaderStatus,
});

const mapDispatchToProps = dispatch => ({
  updateFormStatus(value) {
    dispatch(setFormStatus(value));
  },
  updateStep(step) {
    dispatch(updateStep(step));
  },
  onChange1BODInfo(label, value) {
    dispatch(set1BODInfo({ label, value }));
  },
});

const enhance = compose(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  withStyles(s),
);

export default enhance(FirstBOD);
