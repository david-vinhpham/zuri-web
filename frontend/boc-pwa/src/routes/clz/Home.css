/**
 * Home Credit (https://www.homecredit.vn/)
 *
 * Copyright © 2018-present HomeCredit, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

@import '../../components/variables.css';

:root {
  --container-horizontal-padding: 40px;
  --slogan-height: 36px;
  --logo-width: 60vw;
}

.container {
  margin: 0 auto;
  padding: 0 0 var(--container-horizontal-padding);
  max-width: var(--max-content-width);
}

.logo {
  margin-top: var(--b-space-2xl);
}

.slogan {
  margin-top: var(--b-space-sm);
  margin-bottom: var(--b-space-lg);
  height: var(--slogan-height);
}

.list {
  list-style: none;
  text-align: left;
  padding-left: var(--b-space-lg);
  font-weight: 500;
}

.list li {
  margin-bottom: var(--b-space-sm);
}

.list li::before {
  padding-right: var(--b-space-xs);
  content: '✓';
}

.btnStart {
  border-radius: 0;
}

@media (--xs) {
  :root {
    --slogan-height: 30px;
    --container-background-size: 52%;
  }

  .container {
    padding: 0;
    max-width: unset;
    text-align: center;
    background-color: var(--c-ui-light);
    height: 100%;
    background-image: url('./bg.png');
    background-repeat: no-repeat;
    background-position: center bottom;
    background-size: var(--container-background-size);
  }

  .logo {
    margin-top: var(--b-space-md);
    width: var(--logo-width);
  }

  .slogan {
    margin-bottom: 0;
  }

  .container .bottomBtn {
    text-align: center;
    margin: 0 auto;
    position: absolute;
    bottom: 0;
    width: 100%;
    font-size: var(--f-headline);

    & > button {
      display: inline-block;
    }
  }
}

@media (--xxs) {
  :root {
    --logo-width: 55vw;
  }

  .container {
    height: 100%;
    font-size: var(--font-size-small);
  }

  .slogan {
    margin-top: var(--b-space-xs);
  }

  .list {
    padding-left: var(--b-space-md);
  }

  .list li {
    margin-bottom: var(--b-space-xs);
  }

  .container .bottomBtn {
    font-size: var(--font-size-small);
  }
}
