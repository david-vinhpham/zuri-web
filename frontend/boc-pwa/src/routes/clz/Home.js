/**
 * BOC VN (http://www.bocvietnam.com/)
 *
 * Copyright © 2018-present BOCVN, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import React from 'react';
import withStyles from 'isomorphic-style-loader/lib/withStyles';
import { Button } from 'homecredit-ui';

import history from '../../history';
import slogan from './slogan.svg';
import s from './Home.css';

class Home extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isDirty: false,
    };
    this.trackingBounceRate = this.trackingBounceRate.bind(this);
  }

  componentDidMount() {
    window.ga('send', 'event', {
      eventCategory: 'Welcome Page',
      eventAction: 'Visit_welcome',
      eventLabel: 'Welcome',
    });
    window.addEventListener('beforeunload', this.trackingBounceRate);
  }

  componentWillUnmount() {
    window.removeEventListener('beforeunload', this.trackingBounceRate);
  }

  trackingBounceRate() {
    !this.state.isDirty &&
      window.ga('send', {
        hitType: 'event',
        eventCategory: 'Welcome Page',
        eventAction: 'Bounce_rate',
        eventLabel: 'Welcome',
      });
  }

  handleClick = async () => {
    this.setState({ isDirty: true });
    history.push('/1bod');
  };

  render() {
    return (
      <div className={s.container}>
        <div>
          <img src="/images/small-logo.svg" className={s.logo} alt="Logo" />
          <img src={slogan} className={s.slogan} alt="Slogan" />
          <ul className={s.list}>
            <li>
              Vay tiền mặt lên đến <b>40</b> triệu đồng
            </li>
            <li>
              Lãi suất hấp dẫn chỉ từ <b>1.66%/ tháng</b>
            </li>
            <li>Dành cho khách hàng vay lần đầu tiên</li>
            <li>Dành cho khách hàng từ 20 đến 60 tuổi</li>
            <li>Chỉ cần CMND và Giấy phép lái xe/ Hộ khẩu</li>
            <li>Nhận tiền trong 2 tiếng sau khi ký hợp đồng</li>
          </ul>
          <div className={s.bottomBtn}>
            <Button onClick={this.handleClick} stretch className={s.btnStart}>
              Bắt đầu
            </Button>
          </div>
        </div>
      </div>
    );
  }
}

export default withStyles(s)(Home);
