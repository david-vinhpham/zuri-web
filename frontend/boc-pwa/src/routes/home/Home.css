@import '../../components/variables.css';

.container {
  width: 100%;
  height: 100%;
  position: relative;
  color: var(--c-brand-1);
}

hr {
  border: 1px solid var(--c-brand-4);
  margin: 0;
}

.invoice {
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  padding: var(--size-08) var(--size-14) var(--size-10);
}

.top {
  width: 100%;
  display: flex;
  justify-content: space-between;
}

.header {
  font-size: var(--font-size-16);
  font-weight: var(--font-weight-500);
}

.date {
  font-size: var(--font-size-14);
  font-weight: var(--font-weight-400);
}

.label {
  font-size: var(--font-size-14);
}

.cabinet {
  text-align: center;
  margin: auto var(--size-02);
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
}

.cabinet div:first-child {
  font-weight: var(--font-weight-500);
  font-size: var(--font-size-18);
}

.middle {
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  margin: var(--size-10) auto var(--size-08);
}

.middle span:first-child {
  font-size: var(--size-22);
  font-weight: var(--font-weight-500);
}

.rowTotal {
  width: 100%;
  display: flex;
  margin: var(--size-18) auto auto;
}

.statistic {
  display: flex;
  flex-direction: column;
  justify-content: flex-start;
  align-items: center;
}

.statistic div:first-child {
  font-size: var(--font-size-24);
}

.statistic div:last-child {
  font-size: var(--font-size-14);
}

.bottom {
  display: flex;
  justify-content: space-between;
  box-sizing: border-box;
}

.toolBar {
  width: 100%;
  height: var(--size-56);
  display: flex;
  justify-content: center;
  position: absolute;
  bottom: 0;
  border-top: var(--size-01) solid var(--c-brand-5);
  box-sizing: border-box;
}

.tab {
  width: 25%;
  border-right: var(--size-01) solid var(--c-brand-5);
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  font-size: var(--font-size-12);
  cursor: pointer;
}

.toolBar .tab:last-child {
  border: none;
}
