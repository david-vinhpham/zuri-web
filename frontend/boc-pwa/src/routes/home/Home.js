/**
 * BOC VN (http://www.bocvietnam.com/)
 *
 * Copyright Â© 2019-present BOCVN, LLC. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */
import React from 'react';
import withStyles from 'isomorphic-style-loader/lib/withStyles';

import s from './Home.css';

class Home extends React.Component {
  render() {
    return (
      <div className={s.root}>
        <div className={s.container}>
          <h1>http://www.bocvietnam.com</h1>
        </div>
      </div>
    );
  }
}

export default withStyles(s)(Home);
