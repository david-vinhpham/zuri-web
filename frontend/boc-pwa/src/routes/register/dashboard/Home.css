:root {
  --login-width: 35rem;
  --container-max-w: 600px;
}

.container {
  max-width: var(--container-max-w);
  margin-left: auto;
  margin-right: auto;
  padding: var(--b-space-xl);
  padding-top: var(--b-space-md);
}

.welcome {
  font-weight: var(--font-weight-bold);
  font-size: var(--title-size);
  color: var(--c-text);
  font-style: italic;
}

.numberOfRetry {
  color: var(--c-brand);
}
