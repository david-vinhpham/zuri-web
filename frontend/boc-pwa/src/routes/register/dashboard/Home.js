import React from 'react';
import { compose } from 'redux';
import { connect } from 'react-redux';
import {
  clearUserInfo,
  setLoaderStatus,
  alert,
  registerDevice,
  unRegisterDevice,
  getSAInfo,
  updateUserInfo,
} from 'actions/saRegister';
import { verifyOTP, generateOTP } from 'actions/clz';
import { SA_ALERT } from 'constants/saRegister';
import noop from 'lodash/noop';
import get from 'lodash/get';
import { func, shape, string } from 'prop-types';
import withStyles from 'isomorphic-style-loader/lib/withStyles';

import { isTokenAlive } from '../common/ultis';
import {
  MESSAGE_OVER_RETRY,
  MESSAGE_VERIFY_FAIL,
  MATCHED,
  MESSAGE_RESEND_REMAIN,
  MESSAGE_RETRY_OTP,
  MESSAGE_ALREADY_REGISTER,
  NUMBER_REMAIN,
  MESSAGE_OTP_WRONG,
  MESSAGE_OUT_OF_TIME,
} from './constant';
import history from '../../../history';
import s from './Home.css';
import PhoneInput from './PhoneInput';
import OTP from './OTP';
import Finished from './Finished';

class Home extends React.PureComponent {
  constructor(props) {
    super(props);

    this.state = {
      verifyId: '',
      phoneNumber: '',
      token: '',
    };
  }

  async componentDidMount() {
    const {
      clearUserInfoAction,
      getSAInfoAction,
      setLoaderAction,
    } = this.props;

    // If token is not alive clear localstorage, and push to /login
    if (!isTokenAlive()) {
      clearUserInfoAction();
      history.push('/sa-register/login');
      return;
    }

    // Check if user is registered
    setLoaderAction(true);
    await getSAInfoAction();
    setLoaderAction(false);
  }

  handlePhoneNumberChange = e => {
    this.setState({
      phoneNumber: e.target.value,
    });
  };

  /**
   * Handle send OTP with number
   */
  handleSendOTP = async phoneNumber => {
    const { setLoaderAction } = this.props;

    setLoaderAction(true);
    const result = await generateOTP(phoneNumber);
    setLoaderAction(false);

    return result;
  };

  /**
   * Handle when phone input submit
   */
  handlePhoneInputSubmit = async phoneNumber => {
    const result = await this.handleSendOTP(phoneNumber);
    this.setState(() => ({ phoneNumber, verifyId: result.verificationID }));
  };

  resetState = () => {
    this.setState(() => ({
      verifyId: '',
      phoneNumber: '',
      token: '',
    }));
  };

  /**
   * Handle OTP over retry
   */
  handleOTPOverRetry = () => {
    const { alertAction } = this.props;
    this.resetState();

    alertAction({
      title: SA_ALERT.info.title,
      message: MESSAGE_OVER_RETRY,
    });
  };

  /**
   * Handle OTP retry, show alert and call API
   */
  handleOTPRetry = resendRemain => {
    const { alertAction } = this.props;

    alertAction({
      title: SA_ALERT.info.title,
      message: (
        <p>
          {MESSAGE_RETRY_OTP[0]}{' '}
          <span className={s.numberOfRetry}>{resendRemain}</span>{' '}
          {MESSAGE_RETRY_OTP[1]}
        </p>
      ),
    });
  };

  /**
   * Handle verified OTP,
   * check if retryleft is
   */
  handleVerifyOTP = async (otp, retryLeft) => {
    const { verifyId, phoneNumber } = this.state;
    const { alertAction } = this.props;

    try {
      const response = await verifyOTP(phoneNumber, verifyId, otp);
      const isMatched = get(response, 'data.status') === MATCHED;

      if (isMatched) {
        const token = get(response, 'token.token');
        return token;
      }

      if (retryLeft) {
        alertAction({
          title: SA_ALERT.error.title,
          message: (
            <>
              <p>{MESSAGE_RESEND_REMAIN}</p>
              <p>
                {MESSAGE_OTP_WRONG.resend}{' '}
                <span className={s.numberOfRetry}>
                  {NUMBER_REMAIN(retryLeft)}
                </span>{' '}
                {MESSAGE_OTP_WRONG.times}
              </p>
            </>
          ),
        });
      }

      if (!retryLeft) {
        alertAction({
          title: SA_ALERT.error.title,
          message: (
            <>
              <p>
                {MESSAGE_OUT_OF_TIME.overtry}{' '}
                <span className={s.numberOfRetry}>
                  {MESSAGE_OUT_OF_TIME.maxTime}
                </span>{' '}
                {MESSAGE_OUT_OF_TIME.times}
              </p>
              <p>{MESSAGE_OVER_RETRY}</p>
            </>
          ),
        });

        this.resetState();
      }

      return '';
    } catch (err) {
      alertAction({
        title: SA_ALERT.error.title,
        message: MESSAGE_VERIFY_FAIL,
      });

      return '';
    }
  };

  /**
   * Handle register device
   */
  registerDevice = async (otp, retryLeft) => {
    const {
      setLoaderAction,
      registerDeviceAction,
      auth,
      getSAInfoAction,
      alertAction,
    } = this.props;
    const { phoneNumber, verifyId } = this.state;
    const sessionID = document
      .getElementById('sessionID')
      .getAttribute('value');

    setLoaderAction(true);
    // Verify OTP
    const token = await this.handleVerifyOTP(otp, retryLeft);
    if (!token) {
      setLoaderAction(false);
      return;
    }

    const info = await getSAInfoAction();

    if (get(info, 'createAt')) {
      alertAction({
        title: SA_ALERT.error.title,
        message: MESSAGE_ALREADY_REGISTER,
      });
      setLoaderAction(false);
      return;
    }

    try {
      await registerDeviceAction({
        phoneNumber,
        retailAgent: auth.username,
        sessionID,
        verificationID: verifyId,
      });
      this.setState({ token });
    } catch (err) {
      this.resetState();
    }

    setLoaderAction(false);
  };

  handleUnRegister = async () => {
    const { unRegisterDeviceAction, setLoaderAction } = this.props;

    setLoaderAction(true);
    await unRegisterDeviceAction();
    this.resetState();
    setLoaderAction(false);
  };

  /**
   * Show step get OTP for register device
   */
  renderStepRegister = () => {
    const { verifyId, token, phoneNumber } = this.state;
    const { createAt } = this.props.auth;

    return (
      <React.Fragment>
        {!token && (
          <p className={s.welcome}>
            Chào mừng bạn đến với hệ thống đăng ký thiết bị của BOCVN.
          </p>
        )}

        {!verifyId && <PhoneInput sendOTP={this.handlePhoneInputSubmit} />}

        {verifyId &&
          !createAt && (
            <OTP
              registerDevice={this.registerDevice}
              reSendOTP={this.handleOTPRetry}
              overTry={this.handleOTPOverRetry}
              phoneNumber={phoneNumber}
            />
          )}
      </React.Fragment>
    );
  };

  render() {
    const { createAt } = this.props.auth;

    return (
      <div className={s.container}>
        {!createAt && this.renderStepRegister()}

        {createAt && (
          <Finished createAt={createAt} unRegister={this.handleUnRegister} />
        )}
      </div>
    );
  }
}

Home.defaultProps = {
  clearUserInfoAction: noop,
  setLoaderAction: noop,
  alertAction: noop,
  registerDeviceAction: noop,
  auth: shape({
    createAt: string,
    username: string,
  }),
};

Home.propTypes = {
  clearUserInfoAction: func,
  setLoaderAction: func,
  alertAction: func,
  registerDeviceAction: func,
  auth: shape({
    createAt: '',
    username: '',
  }),
};

const mapDispatchToProps = dispatch => ({
  clearUserInfoAction: () => dispatch(clearUserInfo()),
  setLoaderAction: status => dispatch(setLoaderStatus(status)),
  alertAction: ({ title, message }) => dispatch(alert({ title, message })),
  /**
   * @param {Object} deviceInfo { phoneNumber, retailAgent, sessionID, verificationID }
   */
  registerDeviceAction: deviceInfo => dispatch(registerDevice(deviceInfo)),
  getSAInfoAction: () => dispatch(getSAInfo()),
  updateUserInfoAction: userInfo => dispatch(updateUserInfo(userInfo)),
  unRegisterDeviceAction: () => dispatch(unRegisterDevice()),
});

const mapStateToProps = ({ saRegister }) => ({ auth: saRegister.auth });

const enchane = compose(
  connect(
    mapStateToProps,
    mapDispatchToProps,
  ),
  withStyles(s),
);

export default enchane(Home);
