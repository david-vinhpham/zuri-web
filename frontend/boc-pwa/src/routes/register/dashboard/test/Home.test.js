import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

import Home from '../Home';

const mockStore = configureStore();
const initialState = {
  saRegister: {
    auth: {
      username: null,
    },
  },
};

it('Home snapshot', () => {
  const store = mockStore(initialState);
  const wrapper = renderer.create(
    <Provider store={store}>
      <Home />
    </Provider>,
  );
  const snapshot = wrapper.toJSON();

  expect(snapshot).toMatchSnapshot();
});
