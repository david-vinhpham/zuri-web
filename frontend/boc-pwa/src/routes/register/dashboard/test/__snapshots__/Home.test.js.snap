// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home snapshot 1`] = `
<div
  className="container"
>
  <p
    className="welcome"
  >
    Chào mừng bạn đến với hệ thống đăng ký thiết bị của HomeCredit.
  </p>
  <div>
    <div
      className="Input_wrapper__WALTg"
    >
      <i
        className="material-icons Input_icon__1jjsS"
      >
        
      </i>
      <input
        accept=""
        autoComplete=""
        className="Input_input__1IXXH"
        disabled={false}
        id=""
        maxLength={null}
        name="phoneNumber"
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onMouseDown={[Function]}
        placeholder=""
        readOnly={false}
        type="text"
        value=""
      />
      <label
        className="Input_label__2ZxWx"
        htmlFor=""
      >
        Nhập số điện thoại nhận OTP
      </label>
      <div
        className="Input_underline__Xwqb3"
      />
    </div>
    <button
      className="Button_buttonReset__3cQTl Button_button__2pQRN btnSendOTP"
      disabled={false}
      onClick={[Function]}
      title=""
      type="button"
    >
      Gửi OTP
    </button>
  </div>
  
</div>
`;
